src/git-merge-pr.rs

```diff
@@ -4,18 +4,171 @@ extern crate git_rsl;
+use git_rsl::utils::git;
 
-use workspace::Workspace;
-use git_rsl::{prep_workspace, restore_workspace};
-use git_rsl::utils::git::discover_repo;
 use errors::*;
+use structopt::StructOpt;
+use std::process;
 
-fn main() {
-    let repo = discover_repo().unwrap(); 
-    let ws = Workspace::new(repo).unwrap();
+#[derive(StructOpt, Debug)]
+#[structopt(name = "git merge-pr")]
+struct Opt {
+    /// Keep a local copy of the branch to be merged
+    #[structopt(long="no-prune")]
+    no_prune: bool,
+
+    /// The branch to merge
+    #[structopt(name="src")]
+    branch_to_merge: String,
+
+    /// The destination branch
+    #[structopt(name="dest")]
+    destination_branch: String,
+
+}
+
+fn merge_pr(opts: Opt) -> Result<()> {
+    let mut repo = git::discover_repo()?;
+
+    let current_branch_name = repo.head()?
+        .name()
+        .ok_or("Not on a named branch. Please switch to one so we can put \
+                you back where you started when this is all through.")?
+        .to_owned();
+
+    let local_destination_branch = format!("refs/heads/{}", opts.destination_branch);
+    let local_branch_to_merge = format!("refs/heads/{}", opts.branch_to_merge);
+
+    if !opts.no_prune && current_branch_name == local_branch_to_merge {
+        bail!("You are checked out on the branch you are trying to merge. \
+               This will prevent the tool from auto-pruning after merging. \
+               If you want to keep a local copy of this branch, please consider \
+               the --no-prune flag. Otherwise, check out a different branch \
+               and try again.");
+    }
+
+    let ws = Workspace::new(&mut repo)?;
+
+    // secure-fetch src, dest
+    let mut remote = ws.repo.find_remote("origin")
+        .chain_err(|| "unable to find remote 'origin'")?;
+
+    // secure fetch both branches from origin
+    secure_fetch(&ws.repo,
+                 &mut remote,
+                 &vec![opts.branch_to_merge.as_str(),
+                       opts.destination_branch.as_str()])?;
+
+
+    let origin_destination_branch = format!("refs/remotes/origin/{}", opts.destination_branch);
+    let origin_branch_to_merge = format!("refs/remotes/origin/{}", opts.branch_to_merge);
+
+    let destination_branch_oid = ws.repo.refname_to_id(&origin_destination_branch)?;
+    let branch_to_merge_oid = ws.repo.refname_to_id(&origin_branch_to_merge)?;
 
-    println!("merge-pr");
+    let destination_branch_commit = &ws.repo.find_commit(destination_branch_oid)?;
+    let branch_to_merge_commit = &ws.repo.find_commit(branch_to_merge_oid)?;
+
+    let destination_branch_annotated_commit = &ws.repo.find_annotated_commit(destination_branch_oid)?;
+    let branch_to_merge_annotated_commit = &ws.repo.find_annotated_commit(branch_to_merge_oid)?;
+
+    // The original shell version runs 'git gc' here; I'm omitting it because it
+    // gc is a potentially destructive operation, and doing it automatically is
+    // somewhat dangerous.
```

Comment:
This kind of documentation should probably not live in the code. If it needs to be investigated, then an issue would be good. Otherwise, it would ideally live in a design document.